src/Forms/HTMLEditor/HTMLEditorAttributeRule.php

```diff
@@ -0,0 +1,200 @@
+<?php
+
+namespace SilverStripe\Forms\HTMLEditor;
+
+use DOMAttr;
+use InvalidArgumentException;
+use SilverStripe\Core\ArrayLib;
+
+/**
+ * Rules used to define whether a given attribute or regex pattern of attributes is allowed in the content of an HTMLEditorField.
+ * Each attribute rule is applied to a single HTMLEditorElementRule matching an element or a regex patter of elements.
+ */
+class HTMLEditorAttributeRule
+{
+    public const string VALUE_DEFAULT = 'default';
+
+    public const string VALUE_FORCED = 'forced';
+
+    public const string VALUE_VALID = 'valid';
+
+    /**
+     * Name of the attribute(s) this rule applies to.
+     * This can also be a regex pattern - in which case nameIsPattern should be set to `true`.
+     */
+    private string $name;
+
+    /**
+     * Indicates whether name is a regex pattern to match against.
+     */
+    private bool $nameIsPattern = false;
+
+    /**
+     * A default value which will be used if none is set.
+     * Can only be set in the constructor.
+     */
+    private ?string $defaultValue = null;
+
+    /**
+     * The value the attribute(s) this rule applies to will be forced to, overriding any user-set value.
+     * Can only be set in the constructor.
+     */
+    private ?string $forcedValue = null;
+
+    /**
+     * An allow list of values for the attribute(s) this rule applies to.
+     * If empty, any value is allowed.
+     * Can only be set in the constructor.
+     */
+    private array $validValues = [];
+
+    /**
+     * If true, the attribute(s) this rule applies to must be present for the elements this rule applies to.
+     * Elements which don't have their required attributes are removed.
+     */
+    private bool $isRequired = false;
+
+    /**
+     * Returns a HTMLEditorAttributeRule based on an associative array defining the rule.
+     * If the rule array is empty, the defaults apply.
+     */
+    public static function fromArray(string $name, array $ruleArray): static
```

Comment:
Method should be below the constructor

Reply:
According to https://docs.silverstripe.org/en/5/contributing/coding_conventions/php_coding_conventions/#class-member-ordering static methods come before non-static methods.

Reply:
It's not clear as it doesn't explicitly mention that, or constructor methods in the docs, though given the existing rules we have it would make sense to so that "static is above non-static" in all cases

Reply:
https://github.com/silverstripe/developer-docs/issues/718 with PR to update docs